cabal-testsuite/src/Test/Cabal/TestCode.hs

```diff
@@ -19,22 +27,52 @@ import Data.Typeable     (Typeable)
 data TestCode
     = TestCodeOk
     | TestCodeSkip String
-    | TestCodeKnownFail Int
-    | TestCodeUnexpectedOk
+    | TestCodeKnownFail IssueID
+    | TestCodeUnexpectedOk IssueID
     | TestCodeFail
+    | TestCodeFlakyFailed IssueID
+    | TestCodeFlakyPassed IssueID
   deriving (Eq, Show, Read, Typeable)
 
 instance Exception TestCode
   where
     displayException = displayTestCode
 
 displayTestCode :: TestCode -> String
-displayTestCode TestCodeOk            = "OK"
-displayTestCode (TestCodeSkip msg)    = "SKIP " ++ msg
-displayTestCode (TestCodeKnownFail t) = "OK (known failure, see #" <> show t <> ")"
-displayTestCode TestCodeUnexpectedOk  = "FAIL (unexpected success)"
-displayTestCode TestCodeFail          = "FAIL"
+displayTestCode TestCodeOk               = "OK"
+displayTestCode (TestCodeSkip msg)       = "SKIP " ++ msg
+displayTestCode (TestCodeKnownFail t)    = "OK (known failure, see " <> show t <> ")"
+displayTestCode (TestCodeUnexpectedOk t) = "FAIL (unexpected success, see " <> show t <> ")"
+displayTestCode TestCodeFail             = "FAIL"
+displayTestCode (TestCodeFlakyFailed t)  = "FLAKY (FAIL, see " <> show t <> ")"
+displayTestCode (TestCodeFlakyPassed t)  = "FLAKY (OK, see " <> show t <> ")"
 
 isTestCodeSkip :: TestCode -> Bool
 isTestCodeSkip (TestCodeSkip _) = True
 isTestCodeSkip _                = False
+
+type TestPassed = Bool
+
+newtype IssueID = IssueID Int
+  deriving newtype (Eq, Typeable, Num)
+
+instance Show IssueID where
+  show (IssueID i) = "#" <> show i
```

Comment:
Highjacking `Show` for pretty printing is considered a code smell. Why not a regular function?